File: God Zuo/class028 基数排序/My01.py
import sys

BASE=10
MAXN=100001
arr=[0]*MAXN
helparr=[0]*MAXN
cnts=[0]*BASE
n=0

def radix_sort(bits_count):
    office=1
    while bits_count>0:
        cnts[:]=[0]*BASE
        for i in range(n):
            cnts[(arr[i]//office)%BASE]+=1
        for i in range(1,BASE):
            cnts[i]+=cnts[i-1]
        for i in range (n-1,-1,-1):
             digit=(arr[i]//office)%BASE
             cnts[digit]-=1
             helparr[cnts[digit]]=arr[i]
        arr[:]=helparr[:n]
        office*=BASE
        bits_count-=1

def sort_arr():
    min_val=arr[0]
    for i in range(1,n):
        min_val=min(arr[i],min_val)
    max_val=0
    for i in range(0,n):
        arr[i]-=min_val
        max_val=max(arr[i],max_val)
    radix_sort(bits(max_val))
    for i in range(n):
        arr[i]+=min_val

def bits(num):
    ans=0
    while num>0:
        num//=BASE
        ans+=1
    return ans

data=sys.stdin.read().split()
n=int(data[0])
idx=1
for i in range(n):
    arr[i]=int(data[idx])
    idx+=1
sort_arr()
print(' '.join(str(arr[i])for i in range(n)))

File: God Zuo/class077 区间DP下/My01.cpp
// 完成配对需要的最少字符数量
// 给定一个由'['、']'、'('，')'组成的字符串
// 请问最少插入多少个括号就能使这个字符串的所有括号正确配对
// 例如当前串是 "([[])"，那么插入一个']'即可满足
// 输出最少需要插入多少个字符
// 测试链接 : https://www.nowcoder.com/practice/e391767d80d942d29e6095a935a5b96b

#include<bits/stdc++.h>
using namespace std;

const int maxn = 210;
int dp[maxn][maxn];
string s;

int f(int l,int r){
    if(l==r)
        return 1;
    if(l+1==r){
        if((s[l]=='(' && s[r]==')') || (s[l]=='[' && s[r]==']')){
            return 0;
        }else{
            return 2;
        }
    }
    if(dp[l][r]!=-1)
        return dp[l][r];

    int p1 = INT_MAX;
    if ((s[l] == '(' && s[r] == ')') || (s[l] == '[' && s[r] == ']')){
        p1 = f(l + 1, r - 1);
    }
    int p2 = INT_MAX;
    for (int k = l; k < r;++k){
        p2 = min(p2, f(l, k) + f(k + 1, r));
    }
    int ans = min(p1, p2);
    dp[l][r] = ans;
    return ans;
}

int main(){
    cin >> s;
    int n = s.length();
    for (int i = 0; i < n;++i){
        for (int j = 0; j < n;++j)
            dp[i][j] = -1;
    }
    cout << f(0, n - 1)<<'\n';
    return 0;
}